dll/win32/browseui/CAutoComplete.cpp

```diff
@@ -20,669 +20,2074 @@
-    if (cchText < (size_t)iStart || (INT)cchText <= 0)
-        return;
+CACListView::CACListView() : m_pDropDown(NULL), m_cyItem(CY_ITEM)
+{
+}
 
-    CComHeapPtr<WCHAR> pszText;
-    if (!pszText.Allocate(cchText + 1))
-        return;
+HWND CACListView::Create(HWND hwndParent)
+{
+    ATLASSERT(m_hWnd == NULL);
+    DWORD dwStyle = WS_CHILD | /*WS_VISIBLE |*/ WS_CLIPSIBLINGS | LVS_NOCOLUMNHEADER |
+                    LVS_OWNERDATA | LVS_OWNERDRAWFIXED | LVS_SINGLESEL | LVS_REPORT;
+    HWND hWnd = ::CreateWindowExW(0, GetWndClassName(), L"Internet Explorer", dwStyle,
+                                  0, 0, 0, 0, hwndParent, NULL,
+                                  _AtlBaseModule.GetModuleInstance(), NULL);
+    SubclassWindow(hWnd); // do subclass to handle messages
+    // set extended listview style
+    DWORD exstyle = LVS_EX_ONECLICKACTIVATE | LVS_EX_FULLROWSELECT | LVS_EX_TRACKSELECT;
+    SetExtendedListViewStyle(exstyle, exstyle);
+    // insert one column (needed to insert items)
+    LV_COLUMNW column = { LVCF_FMT | LVCF_WIDTH };
+    column.fmt = LVCFMT_LEFT;
+    column.cx = CX_LIST - ::GetSystemMetrics(SM_CXVSCROLL);
+    InsertColumn(0, &column);
+    return m_hWnd;
+}
 
-    if (GetWindowTextW(hwndEdit, pszText, cchText + 1) <= 0)
-        return;
+// set font handle
+VOID CACListView::SetFont(HFONT hFont)
+{
+    SendMessageW(WM_SETFONT, (WPARAM)hFont, TRUE); // set font
 
-    WORD types[2];
-    for (--iStart; 0 < iStart; --iStart)
+    // get listview item height
+    m_cyItem = CY_ITEM;
+    HDC hDC = GetDC();
+    if (hDC)
     {
-        GetStringTypeW(CT_CTYPE1, &pszText[iStart - 1], 2, types);
-        if (((types[0] & C1_PUNCT) && !(types[1] & C1_SPACE)) ||
-            ((types[0] & C1_SPACE) && (types[1] & (C1_ALPHA | C1_DIGIT))))
+        HGDIOBJ hFontOld = ::SelectObject(hDC, hFont);
+        TEXTMETRICW tm;
+        if (::GetTextMetricsW(hDC, &tm))
         {
-            SendMessageW(hwndEdit, EM_SETSEL, iStart, iEnd);
-            SendMessageW(hwndEdit, EM_REPLACESEL, TRUE, (LPARAM)L"");
-            return;
+            m_cyItem = (tm.tmHeight * 3) / 2; // 3/2 of text height
         }
+        ::SelectObject(hDC, hFontOld);
+        ReleaseDC(hDC);
     }
+}
 
-    if (iStart == 0)
-    {
-        SendMessageW(hwndEdit, EM_SETSEL, iStart, iEnd);
-        SendMessageW(hwndEdit, EM_REPLACESEL, TRUE, (LPARAM)L"");
-    }
+// get the number of visible items
+INT CACListView::GetVisibleCount()
+{
+    if (m_cyItem <= 0) // avoid "division by zero"
+        return 0;
+    CRect rc;
+    GetClientRect(&rc);
+    return rc.Height() / m_cyItem;
 }
 
-/*
-  Window procedure for autocompletion
- */
-LRESULT APIENTRY CAutoComplete::ACEditSubclassProc(HWND hwnd, UINT uMsg, WPARAM wParam, LPARAM lParam)
+// get the text of an item
+CStringW CACListView::GetItemText(INT iItem)
 {
-    CAutoComplete *pThis = static_cast<CAutoComplete *>(GetPropW(hwnd, autocomplete_propertyW));
```

Comment:
Windows has this SetPropW / GetPropW with prop title "CAutoComplete_This" name for keeping the context.

Reply:
However it's no use because ATL can associate HWND and pointer.

Reply:
GetProp/SetProp is slow.